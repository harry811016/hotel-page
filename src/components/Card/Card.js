import React from 'react'
import styled from '@emotion/styled'

const CardContainer = styled.div`
  padding: 7rem 0 5rem;
  display: flex;
  flex-flow: column;
  justify-content: flex-start;
  align-items: center;
`

const CardItem = styled.div`
  width: 850px;
  height: 250px;
  border-radius: 8px;
  border: 1px solid transparent;
  box-shadow: 0 0 2px 0 rgba(19,26,31,0.12),0 2px 4px 0 rgba(19,26,31,0.22);
  margin-bottom: 2rem;
  display: flex;
  position: relative;

  .img{
    img{
      width: 250px;
      height: 250px;
      border-bottom-left-radius: 8px;
      border-top-left-radius: 8px;
      object-fit: cover;
      object-position: 50% 50%;
    }
  }

  .description {
    height: 250px;
    width: calc(106%);
    top: 0;
    left: 0;
    position: absolute;
    padding: 0.75rem;
    background-color: ${props => props.theme.colors.white};
    z-index: 200;
    overflow-y: auto;

    &__text {
      flex: 0 0 95%;
      font-size: 0.75rem;
      padding-right: 0.25rem;
      font-weight: 400;
    }

    &__button {
      flex: 0 0 5%;
      height: 1.5rem;
      line-height: 1.5rem;
      font-size: 0.8rem;
      color: ${props => props.theme.colors.dark_blue};
      border-radius: 8px;
      margin: 0 auto;
      padding-left: 0.25rem;
      font-weight: 1000px;
      text-align: center;
      cursor: pointer;
    }
  }

  .info {
    padding: 1rem 2rem;
    width: 400px;
    display: flex;
    flex-flow: column;
    justify-content: space-between;
    position: relative;

    &__main {
      &__name {
        font-size: 1.5rem;
        font-weight: 600;
      }

      &__address {
        font-size: 0.75em;
        padding: 0.5rem 0;
        word-wrap: normal;
      }

      &__stars {
        color:  ${props => props.theme.colors.dark_yellow};
      }
    }

    &__bottomWrapper {
      display: flex;
      justify-content: space-between;
      align-items: center;
      margin-top: auto;
      margin-bottom: 0.5rem;

      &__rating {
        color: #ffffff;
        width: 30px;
        height: 30px;
        line-height: 30px;
        font-weight: 600;
        text-align: center;
        background-color: ${props => props.theme.colors.dark_blue};
        border-radius: 8px;
      }

      &__more {
        display: flex;
        align-items: flex-end;
        height: 100%;
        color:  ${props => props.theme.colors.dark_blue};
        font-size: 0.8rem;
        font-style: italic;
        cursor: pointer;
      }
    }

    &__competitors{
      width: 100%;
      align-items: flex-end;
      display: flex;
      justify-content: flex-start;
      overflow-y: hidden;
      overflow-x: auto;
      border-top: 1px solid ${props => props.theme.colors.bg_grey};
      padding-top: 10px;

      &__item {
        height: 45px;
        padding: 0 1rem;
        font-size: 0.7rem;
        line-height: 0.7rem;
        display: flex;
        flex-flow: column;
        justify-content: center;
        align-items: center;
        white-space: nowrap;

        .name {
          font-weight: 800;
          margin-bottom: -0.5rem;
        }
      }

      &__item.highlight {
        border-radius: 5px;
        color: ${props => props.theme.colors.dark_green};
        background-color: ${props => props.theme.colors.bg_green};
        }
    }
  }

  .deal{
    display: flex;
    flex-flow: column-reverse;
    margin: 1rem auto;
    position: relative;
    border-left: 1px solid ${props => props.theme.colors.bg_grey};
    padding-left: 1rem;

    &__price {
      font-weight: 750;
      font-size: 1.25rem;

      &__tax {
        display: flex;
        flex-flow: column;
        justify-content: flex-start;
        height: 15px;
        font-weight: 400;
        font-size: 0.8rem;
        line-height:  0.8rem;
        margin: -0.5rem 0 1rem;

        p {
          border-bottom: 1px dotted ${props => props.theme.colors.black};
        }
      }
    }

    &__button {
      color:  ${props => props.theme.colors.white};
      background-color:  ${props => props.theme.colors.dark_grey};
      border-radius: 2px;
      width: 120px;
      height: 35px;
      line-height: 35px;
      font-weight: 600;
      text-align: center;
      margin-top: 0.6rem;
      box-shadow: 0 0 2px 0 rgba(19,26,31,0.12),0 2px 4px 0 rgba(19,26,31,0.22);
    }

    &__button.cursor{
      cursor: pointer;
      background-color: ${props => props.theme.colors.main_color};
      &:hover {
        transform: scale(1.05);
      }
    }

    &__savedCost {
      color: ${props => props.theme.colors.dark_red};
      font-size: 0.8rem;
      line-height: 0.8rem;
      margin: 0.5rem 0;

      &--highlight {
        font-weight: 800;
      }

      &--strikethrough {
        font-style: italic;
        text-decoration: line-through;
        font-weight: 400;
        margin-right: 1rem;
      }
    }
  }

  .tooltip {
    position: relative;
    display: inline-block;
    cursor: pointer;
  }

  .tooltip {
    .tooltiptext {
      visibility: hidden;
      display: flex;
      flex-flow: column;
      justify-content: flex-start;
      align-items: flex-start;
      width: 150px;
      background-color: ${props => props.theme.colors.white};
      border: 1px solid ${props => props.theme.colors.dark_yellow};
      border-radius: 5px;
      text-align: center;
      border-radius: 6px;
      padding: 0.5rem;
      font-size: 0.8rem;
      
      position: absolute;
      transform: translate(-105%, -100%);
      z-index: 1;

      p {
        line-height: 0px;
        font-weight: 600;
      }
    }
  }

  .tooltip:hover .tooltiptext {
    visibility: visible;
  }
`

const Card = (props) => {
  const hoteldata = props.hoteldata
  const starConverter = (stars) => {
    stars = Math.round(stars)
    let str = ''
    for (let i = 0; i < stars; i++) {
      str += 'â˜…'
    }
    return str
  }

  const priceConverter = (currency, price) => {
    price = Math.round(price)

    if (['KRW', 'JPY', 'IDR'].includes(currency,)) {
      price = price.toLocaleString()
    }

    return `${currency} ${price}`
  }

  const showDescription = (e) => {
    const id = Number(e.target.id)
    const modal = e.target.nextSibling
    const [text, closeBtn] = e.target.nextSibling.children

    // insert HTML content
    let data = hoteldata.filter(item => item.id === id)
    data = data[0].description.replace(/<br>/g, '')
    text.innerHTML = data

    // open modal
    modal.classList.remove('d-none')

    // close modal
    closeBtn.addEventListener('click', () => {
      modal.scrollTop = 0
      text.innerHTML = ''
      modal.classList.add('d-none')
    })
  }

  return (
    <CardContainer>
      {hoteldata.map(item => {
        return (
          <CardItem key={item.id}>
            {/* Left part - pciture */}
            <div className="img"><img src={item.photo} alt="" /></div>

            {/* Middle part - hotel and competitor info */}
            <div className="info">
              <div className="info__main">
                <div className="info__main__name" data-testid="name">{item.name}</div>
                <div className="info__main__address" data-testid="address">{item.address}</div>
                {item.stars || item.stars === 0 ? (<div className="info__main__stars" data-testid="stars">{starConverter(item.stars)}</div>) : ''}
              </div>
              <div className="info__bottomWrapper">
                {item.rating || item.rating === 0 ? (<div className="info__bottomWrapper__rating" data-testid="rating">{item.rating}</div>) : ''}
                <div className="info__bottomWrapper__more" onClick={showDescription} id={item.id} data-testid="more">see more</div>

                {/* description */}
                <div className="description d-none" >
                  <div className="description__text" data-testid="description">
                  </div>
                  <div className="description__button" data-testid="description-btn">back</div>
                </div>
              </div>

              {/* competitors */}
              <div className="info__competitors">
                {item.price && item.price.competitionSet.length > 1 && item.price.savedCost > 0 ? (item.price.competitionSet.map(item => {
                  if (item.name === 'Our Price') {
                    return (<div className="info__competitors__item highlight">
                      <p className='name'>{item.name}</p>
                      <p className='price'>{Math.round(item.price).toLocaleString()}</p>
                    </div>)
                  } else {
                    return (<div className="info__competitors__item">
                      <p className='name'>{item.name}</p>
                      <p className='price'>{Math.round(item.price).toLocaleString()}</p>
                    </div>)
                  }
                })) : ''}
              </div>
            </div>

            {/* Right part - price and CTA button */}
            <div className="deal" data-testid="deal">
              {item.price ? (
                <div className="deal__button cursor" data-testid="cta-btn">
                  Book now!
                </div>) : (
                  <div className="deal__button">
                    Unavailable
                  </div>)}

              <div className="deal__price">
                <div className="tooltip" >
                  {item.price ? (
                    <span data-testid="price">{priceConverter(props.currency, item.price.price)}</span>
                  ) : ''}

                  {item.price && item.price.taxes_and_fees ? (<div className="deal__price__tax" data-testid="tax-included-msg">
                    <p>*tax-inclusive</p>
                  </div>) : ''}
                  {item.price && item.price.taxes_and_fees ? (
                    <span className="tooltiptext">
                      <p className='fees' data-testid="tax">{item.price.taxes_and_fees.tax ? `tax: $${item.price.taxes_and_fees.tax}` : ''}</p>
                      <p className='fees' data-testid="hotel-fee">{item.price.taxes_and_fees.hotel_fees ? `hotel fees: $${item.price.taxes_and_fees.hotel_fees}` : ''}</p>
                    </span>) : ''}
                </div>
              </div>

              {item.price && item.price.savedCost > 0 ? (
                <div className="deal__savedCost">
                  <span className="deal__savedCost--strikethrough" data-testid="expensive-price">${Math.round(item.price.savedCost + item.price.price).toLocaleString()}</span>
                  <span className="deal__savedCost--highlight" data-testid="saved-percentage">Save {(100 * item.price.savedCost / item.price.price).toFixed(1)}% !</span>
                </div>) : ''}

            </div>
          </CardItem>
        )
      })
      }
    </CardContainer>
  )
}

export default Card