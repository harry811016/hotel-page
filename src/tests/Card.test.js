import React from 'react'
import { render, cleanup, fireEvent } from '@testing-library/react'
import { ThemeProvider } from 'emotion-theming'
import Card from '../components/Card/Card'
import mockDataSet from '../tests/mockDataSet'

const theme = {
  colors: {
    main_color: "#ff690f",
    black: "#000000",
    white: "#ffffff",
    dark_blue: "#003580",
    dark_yellow: "#feba02",
    dark_green: "#008009",
    dark_grey: "#333333",
    dark_red: "#bf323b",
    bg_green: "rgba(0,128,9,0.05)",
    bg_grey: "rgba(200, 200, 200, 0.5)"
  }
}

afterEach(cleanup)

// snapshot
it('should take a snapshot', () => {
  const { asFragment } = render(
    <ThemeProvider theme={theme}>
      <Card
        currency={mockDataSet.set1.currency}
        hoteldata={mockDataSet.set1.data}
      />
    </ThemeProvider>)
  expect(asFragment(<ThemeProvider theme={theme}><Card /></ThemeProvider>)).toMatchSnapshot()
})

// [Mock data set 1] show basic info with SGD
it('select currency as SGD', () => {
  const currency = mockDataSet.set1.currency
  const data = mockDataSet.set1.data[0]
  const { getByTestId } = render(
    <ThemeProvider theme={theme}>
      <Card
        currency={mockDataSet.set1.currency}
        hoteldata={mockDataSet.set1.data}
      />
    </ThemeProvider>)

  expect(getByTestId('name')).toHaveTextContent('hotel name')
  expect(getByTestId('address')).toHaveTextContent('hotel address')
  expect(getByTestId('stars')).toHaveTextContent('★★★★')
  expect(getByTestId('rating')).toHaveTextContent('7.7')

  // no description before clicking "see more" btn
  expect(getByTestId('description')).toHaveTextContent('')

  expect(getByTestId('price')).toHaveTextContent('SGD 100')
  expect(getByTestId('expensive-price')).toHaveTextContent('$105')
  expect(getByTestId('saved-percentage')).toHaveTextContent('Save 5.0% !')
  expect(getByTestId('deal')).toHaveTextContent('Book now!')

  // tax and hotel fee exist if there are data
  expect(getByTestId('tax-included-msg')).toHaveTextContent('*tax-inclusive')
  expect(getByTestId('tax')).toHaveTextContent('tax: $13.12')
  expect(getByTestId('hotel-fee')).toHaveTextContent('hotel fees: $16.4')
})

// show description after clicked see more button
it('show description after clicked "see more" button', () => {
  const data = mockDataSet.set1.data[0]
  const { getByTestId } = render(
    <ThemeProvider theme={theme}>
      <Card
        currency={mockDataSet.set1.currency}
        hoteldata={mockDataSet.set1.data}
      />
    </ThemeProvider>)
  fireEvent.click(getByTestId('more'))
  expect(getByTestId('description')).toHaveTextContent('hotel description')
  fireEvent.click(getByTestId('description-btn'))
  expect(getByTestId('description')).not.toHaveTextContent('hotel description')
})

// [Mock data set 2] show basic info with KRW
it('change currency to KRW', () => {
  const { getByTestId, queryByText } = render(
    <ThemeProvider theme={theme}>
      <Card
        currency={mockDataSet.set2.currency}
        hoteldata={mockDataSet.set2.data}
      />
    </ThemeProvider>)

  expect(getByTestId('name')).toHaveTextContent('KRW hotel name')
  expect(getByTestId('address')).toHaveTextContent('KRW hotel address')
  expect(getByTestId('stars')).toHaveTextContent('★')
  expect(getByTestId('rating')).toHaveTextContent('1.7')
  expect(getByTestId('saved-percentage')).toHaveTextContent('Save 11.0% !')
  expect(getByTestId('deal')).toHaveTextContent('Book now!')

  // no description before clicking "see more" btn
  expect(getByTestId('description')).toHaveTextContent('')

  // round up price
  expect(getByTestId('price')).toHaveTextContent('KRW 134,435')
  expect(getByTestId('expensive-price')).toHaveTextContent('$149,251')

  // things should not exist
  const taxMsgElement = queryByText('*tax-inclusive')
  expect(taxMsgElement).toBeNull()
  const taxElement = queryByText('tax:')
  expect(taxElement).toBeNull()
  const hotelFeeElement = queryByText('hotel fees:')
  expect(hotelFeeElement).toBeNull()
})

// [Mock data set 3] no price data
it('price data unavailable', () => {
  const { getByTestId, queryByText } = render(
    <ThemeProvider theme={theme}>
      <Card
        currency={mockDataSet.set3.currency}
        hoteldata={mockDataSet.set3.data}
      />
    </ThemeProvider>)
  expect(getByTestId('deal')).toHaveTextContent('Unavailable')
})

// [Mock data set 4] star and rating are zero
it('price data unavailable', () => {
  const { getByTestId } = render(
    <ThemeProvider theme={theme}>
      <Card
        currency={mockDataSet.set4.currency}
        hoteldata={mockDataSet.set4.data}
      />
    </ThemeProvider>)
  expect(getByTestId('stars')).toHaveTextContent('')
  expect(getByTestId('rating')).toHaveTextContent('0')
})