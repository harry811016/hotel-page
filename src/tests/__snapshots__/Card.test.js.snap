// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-t9jjiw"
  >
    <div
      class="css-w4szxx"
    >
      <div
        class="img"
      >
        <img
          alt=""
          src="https://d2ey9sqrvkqdfs.cloudfront.net/ZqSQ/i1_t.jpg"
        />
      </div>
      <div
        class="info"
      >
        <div
          class="info__main"
        >
          <div
            class="info__main__name"
            data-testid="name"
          >
            hotel name
          </div>
          <div
            class="info__main__address"
            data-testid="address"
          >
            hotel address
          </div>
          <div
            class="info__main__stars"
            data-testid="stars"
          >
            ★★★★
          </div>
        </div>
        <div
          class="info__bottomWrapper"
        >
          <div
            class="info__bottomWrapper__rating"
            data-testid="rating"
          >
            7.7
          </div>
          <div
            class="info__bottomWrapper__more"
            data-testid="more"
            id="1"
          >
            see more
          </div>
          <div
            class="description d-none"
          >
            <div
              class="description__text"
              data-testid="description"
            />
            <div
              class="description__button"
              data-testid="description-btn"
            >
              back
            </div>
          </div>
        </div>
        <div
          class="info__competitors"
        >
          <div
            class="info__competitors__item highlight"
          >
            <p
              class="name"
            >
              Our Price
            </p>
            <p
              class="price"
            >
              100
            </p>
          </div>
          <div
            class="info__competitors__item"
          >
            <p
              class="name"
            >
              Booking.com
            </p>
            <p
              class="price"
            >
              105
            </p>
          </div>
        </div>
      </div>
      <div
        class="deal"
        data-testid="deal"
      >
        <div
          class="deal__button cursor"
          data-testid="cta-btn"
        >
          Book now!
        </div>
        <div
          class="deal__price"
        >
          <div
            class="tooltip"
          >
            <span
              data-testid="price"
            >
              SGD 100
            </span>
            <div
              class="deal__price__tax"
              data-testid="tax-included-msg"
            >
              <p>
                *tax-inclusive
              </p>
            </div>
            <span
              class="tooltiptext"
            >
              <p
                class="fees"
                data-testid="tax"
              >
                tax: $13.12
              </p>
              <p
                class="fees"
                data-testid="hotel-fee"
              >
                hotel fees: $16.4
              </p>
            </span>
          </div>
        </div>
        <div
          class="deal__savedCost"
        >
          <span
            class="deal__savedCost--strikethrough"
            data-testid="expensive-price"
          >
            $105
          </span>
          <span
            class="deal__savedCost--highlight"
            data-testid="saved-percentage"
          >
            Save 5.0% !
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
